File: src/dev_health_ops/api/admin/routers/audit_logs.py
from __future__ import annotations

import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from dev_health_ops.api.admin.middleware import get_admin_org_id, require_superuser
from dev_health_ops.api.admin.schemas import AuditLogListResponse, AuditLogResponse
from dev_health_ops.api.services.audit import AuditLogFilter as ServiceAuditLogFilter
from dev_health_ops.api.services.audit import AuditService
from dev_health_ops.api.services.auth import AuthenticatedUser
from dev_health_ops.licensing import require_feature
from dev_health_ops.models.audit import AuditLog

from .common import get_session

router = APIRouter()


def _audit_log_response(log: object) -> AuditLogResponse:
    return AuditLogResponse.model_validate(
        {
            "id": str(getattr(log, "id")),
            "org_id": str(getattr(log, "org_id")),
            "user_id": (
                str(getattr(log, "user_id"))
                if getattr(log, "user_id") is not None
                else None
            ),
            "action": str(getattr(log, "action")),
            "resource_type": str(getattr(log, "resource_type")),
            "resource_id": str(getattr(log, "resource_id")),
            "description": getattr(log, "description"),
            "changes": getattr(log, "changes"),
            "request_metadata": getattr(log, "request_metadata"),
            "status": str(getattr(log, "status")),
            "error_message": getattr(log, "error_message"),
            "created_at": getattr(log, "created_at"),
        }
    )


@router.get("/audit-logs", response_model=AuditLogListResponse)
@require_feature("audit_log", required_tier="enterprise")
async def list_audit_logs(
    session: AsyncSession = Depends(get_session),
    org_id: str = Depends(get_admin_org_id),
    user_id: str | None = Query(None, description="Filter by user ID"),
    action: str | None = Query(None, description="Filter by action type"),
    resource_type: str | None = Query(None, description="Filter by resource type"),
    resource_id: str | None = Query(None, description="Filter by resource ID"),
    status: str | None = Query(None, description="Filter by status (success/failure)"),
    start_date: datetime | None = Query(
        None, description="Filter logs after this date"
    ),
    end_date: datetime | None = Query(None, description="Filter logs before this date"),
    limit: int = Query(50, ge=1, le=500, description="Maximum number of results"),
    offset: int = Query(0, ge=0, description="Offset for pagination"),
) -> AuditLogListResponse:
    """List audit logs for the organization with optional filters.

    Requires Enterprise tier (audit_log feature).
    """
    svc = AuditService(session)

    filters = ServiceAuditLogFilter(
        user_id=user_id,
        action=action,
        resource_type=resource_type,
        resource_id=resource_id,
        status=status,
        start_date=start_date,
        end_date=end_date,
    )

    logs, total = await svc.get_logs(
        org_id=uuid.UUID(org_id),
        filters=filters,
        limit=limit,
        offset=offset,
    )

    return AuditLogListResponse(
        items=[_audit_log_response(log) for log in logs],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/platform/audit-logs", response_model=AuditLogListResponse)
async def list_platform_audit_logs(
    session: AsyncSession = Depends(get_session),
    current_user: AuthenticatedUser = Depends(require_superuser),
    user_id: str | None = Query(None, description="Filter by user ID"),
    action: str | None = Query(None, description="Filter by action type"),
    resource_type: str | None = Query(None, description="Filter by resource type"),
    resource_id: str | None = Query(None, description="Filter by resource ID"),
    status: str | None = Query(None, description="Filter by status (success/failure)"),
    start_date: datetime | None = Query(
        None, description="Filter logs after this date"
    ),
    end_date: datetime | None = Query(None, description="Filter logs before this date"),
    limit: int = Query(50, ge=1, le=500, description="Maximum number of results"),
    offset: int = Query(0, ge=0, description="Offset for pagination"),
) -> AuditLogListResponse:
    conditions = []
    if user_id:
        conditions.append(AuditLog.user_id == uuid.UUID(user_id))
    if action:
        conditions.append(AuditLog.action == action)
    if resource_type:
        conditions.append(AuditLog.resource_type == resource_type)
    if resource_id:
        conditions.append(AuditLog.resource_id == resource_id)
    if status:
        conditions.append(AuditLog.status == status)
    if start_date:
        conditions.append(AuditLog.created_at >= start_date)
    if end_date:
        conditions.append(AuditLog.created_at <= end_date)

    count_stmt = select(func.count()).select_from(AuditLog)
    if conditions:
        count_stmt = count_stmt.where(*conditions)
    total = int((await session.execute(count_stmt)).scalar_one())

    created_at_col = getattr(AuditLog, "created_at")
    logs_stmt = (
        select(AuditLog).order_by(created_at_col.desc()).limit(limit).offset(offset)
    )
    if conditions:
        logs_stmt = logs_stmt.where(*conditions)
    logs_result = await session.execute(logs_stmt)
    logs = logs_result.scalars().all()

    return AuditLogListResponse(
        items=[_audit_log_response(log) for log in logs],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/audit-logs/{log_id}", response_model=AuditLogResponse)
@require_feature("audit_log", required_tier="enterprise")
async def get_audit_log(
    log_id: str,
    session: AsyncSession = Depends(get_session),
    org_id: str = Depends(get_admin_org_id),
) -> AuditLogResponse:
    """Get a specific audit log entry by ID.

    Requires Enterprise tier (audit_log feature).
    """
    svc = AuditService(session)

    log = await svc.get_log_by_id(
        org_id=uuid.UUID(org_id),
        log_id=uuid.UUID(log_id),
    )

    if not log:
        raise HTTPException(status_code=404, detail="Audit log not found")

    return _audit_log_response(log)


@router.get(
    "/audit-logs/resource/{resource_type}/{resource_id}",
    response_model=list[AuditLogResponse],
)
@require_feature("audit_log", required_tier="enterprise")
async def get_resource_audit_history(
    resource_type: str,
    resource_id: str,
    session: AsyncSession = Depends(get_session),
    org_id: str = Depends(get_admin_org_id),
    limit: int = Query(50, ge=1, le=500, description="Maximum number of results"),
) -> list[AuditLogResponse]:
    """Get audit history for a specific resource.

    Requires Enterprise tier (audit_log feature).
    """
    svc = AuditService(session)

    logs = await svc.get_resource_history(
        org_id=uuid.UUID(org_id),
        resource_type=resource_type,
        resource_id=resource_id,
        limit=limit,
    )

    return [_audit_log_response(log) for log in logs]


@router.get("/audit-logs/user/{user_id}", response_model=list[AuditLogResponse])
@require_feature("audit_log", required_tier="enterprise")
async def get_user_audit_activity(
    user_id: str,
    session: AsyncSession = Depends(get_session),
    org_id: str = Depends(get_admin_org_id),
    limit: int = Query(50, ge=1, le=500, description="Maximum number of results"),
) -> list[AuditLogResponse]:
    """Get audit log activity for a specific user.

    Requires Enterprise tier (audit_log feature).
    """
    svc = AuditService(session)

    logs = await svc.get_user_activity(
        org_id=uuid.UUID(org_id),
        user_id=uuid.UUID(user_id),
        limit=limit,
    )

    return [_audit_log_response(log) for log in logs]

File: internal/apiservice/admin/auditlogs.go
package admin

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"

	"github.com/full-chaos/dev-health-ops/internal/api/policy"
	"github.com/full-chaos/dev-health-ops/internal/api/pybody"
	"github.com/full-chaos/dev-health-ops/internal/api/pyjson"
	"github.com/full-chaos/dev-health-ops/internal/auth/httpapi"
	"github.com/full-chaos/dev-health-ops/internal/pythonparity"
)

const auditLogFeature = "audit_log"

func (h *handlers) auditLogRoutes() []httpapi.Route {
	return []httpapi.Route{
		{Method: http.MethodGet, Pattern: governancePrefix + "/audit-logs", Handler: h.guard.Wrap(policy.Admin, http.HandlerFunc(h.listAuditLogs))},
		{Method: http.MethodGet, Pattern: governancePrefix + "/platform/audit-logs", Handler: h.guard.Wrap(policy.Superuser, http.HandlerFunc(h.listPlatformAuditLogs))},
		{Method: http.MethodGet, Pattern: governancePrefix + "/audit-logs/{log_id}", Handler: h.guard.Wrap(policy.Admin, http.HandlerFunc(h.getAuditLog))},
		{Method: http.MethodGet, Pattern: governancePrefix + "/audit-logs/resource/{resource_type}/{resource_id}", Handler: h.guard.Wrap(policy.Admin, http.HandlerFunc(h.getResourceAuditHistory))},
		{Method: http.MethodGet, Pattern: governancePrefix + "/audit-logs/user/{user_id}", Handler: h.guard.Wrap(policy.Admin, http.HandlerFunc(h.getUserAuditActivity))},
	}
}

type auditLog struct {
	ID, OrgID                        uuid.UUID
	UserID                           *uuid.UUID
	Action, ResourceType, ResourceID string
	Description                      *string
	Changes, RequestMetadata         []byte
	Status                           string
	ErrorMessage                     *string
	CreatedAt                        time.Time
}

const auditLogColumns = `id, org_id, user_id, action, resource_type, resource_id, description, changes, request_metadata, status, error_message, created_at`

func scanAuditLog(row pgx.Row) (*auditLog, error) {
	var log auditLog
	err := row.Scan(&log.ID, &log.OrgID, &log.UserID, &log.Action, &log.ResourceType, &log.ResourceID,
		&log.Description, &log.Changes, &log.RequestMetadata, &log.Status, &log.ErrorMessage, &log.CreatedAt)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &log, nil
}

// auditLogObject is audit_logs.py's _audit_log_response: changes and
// request_metadata must be an object or null (dict[str, Any] | None), or
// response validation fails (a 500).
func auditLogObject(log *auditLog) (*pyjson.Object, error) {
	changes, ok, err := jsonColumnObject(log.Changes)
	if err != nil {
		return nil, err
	}
	if !ok {
		return nil, errors.New("audit log changes is not an object")
	}
	metadata, ok, err := jsonColumnObject(log.RequestMetadata)
	if err != nil {
		return nil, err
	}
	if !ok {
		return nil, errors.New("audit log request_metadata is not an object")
	}
	out := pyjson.NewObject()
	out.Set("id", log.ID.String())
	out.Set("org_id", log.OrgID.String())
	out.Set("user_id", uuidOrNil(log.UserID))
	out.Set("action", log.Action)
	out.Set("resource_type", log.ResourceType)
	out.Set("resource_id", log.ResourceID)
	out.Set("description", optionalString(log.Description))
	out.Set("changes", changes)
	out.Set("request_metadata", metadata)
	out.Set("status", log.Status)
	out.Set("error_message", optionalString(log.ErrorMessage))
	out.Set("created_at", pyTimeString(log.CreatedAt))
	return out, nil
}

func auditLogList(logs []*auditLog) ([]pyjson.Value, error) {
	list := make([]pyjson.Value, 0, len(logs))
	for _, log := range logs {
		object, err := auditLogObject(log)
		if err != nil {
			return nil, err
		}
		list = append(list, object)
	}
	return list, nil
}

// auditFilter is AuditLogFilter: every field is applied only when truthy.
type auditFilter struct {
	userID, action, resourceType, resourceID, status string
	start, end                                       *time.Time
}

// auditListQuery validates the list routes' shared query parameters in
// declaration order and returns the filter and page.
func auditListQuery(r *http.Request) (auditFilter, int64, int64, pybody.Errors) {
	values := r.URL.Query()
	var errs pybody.Errors
	filter := auditFilter{
		userID: truthyQuery(values, "user_id"), action: truthyQuery(values, "action"),
		resourceType: truthyQuery(values, "resource_type"), resourceID: truthyQuery(values, "resource_id"),
		status: truthyQuery(values, "status"),
	}
	if parsed, ok := errs.QueryDatetime("start_date", queryPtr(values, "start_date")); ok && parsed != nil {
		instant := parsed.Time.UTC()
		filter.start = &instant
	}
	if parsed, ok := errs.QueryDatetime("end_date", queryPtr(values, "end_date")); ok && parsed != nil {
		instant := parsed.Time.UTC()
		filter.end = &instant
	}
	limit, offset := pageParams(&errs, values, 50, 500)
	return filter, limit, offset, errs
}

// where renders the filter's conditions after any leading ones. A malformed
// user_id is `uuid.UUID(...)` raising, an unhandled 500.
func (f auditFilter) where(conditions []string, args []any) ([]string, []any, bool) {
	add := func(column string, value any) {
		args = append(args, value)
		conditions = append(conditions, fmt.Sprintf("%s $%d", column, len(args)))
	}
	if f.userID != "" {
		id, err := pythonparity.ParseUUID(f.userID)
		if err != nil {
			return nil, nil, false
		}
		add("user_id =", id)
	}
	if f.action != "" {
		add("action =", f.action)
	}
	if f.resourceType != "" {
		add("resource_type =", f.resourceType)
	}
	if f.resourceID != "" {
		add("resource_id =", f.resourceID)
	}
	if f.status != "" {
		add("status =", f.status)
	}
	if f.start != nil {
		add("created_at >=", *f.start)
	}
	if f.end != nil {
		add("created_at <=", *f.end)
	}
	return conditions, args, true
}

func (s pgStore) auditPage(ctx context.Context, conditions []string, args []any, limit, offset int64) ([]*auditLog, int64, error) {
	where := ""
	if len(conditions) > 0 {
		where = " WHERE " + strings.Join(conditions, " AND ")
	}
	var total int64
	if err := s.Pool.QueryRow(ctx, `SELECT count(*) FROM audit_logs`+where, args...).Scan(&total); err != nil {
		return nil, 0, err
	}
	pageArgs := append(append([]any{}, args...), limit, offset)
	logs, err := s.queryAuditLogs(ctx, `SELECT `+auditLogColumns+` FROM audit_logs`+where+
		fmt.Sprintf(` ORDER BY created_at DESC LIMIT $%d OFFSET $%d`, len(args)+1, len(args)+2), pageArgs...)
	return logs, total, err
}

func (s pgStore) queryAuditLogs(ctx context.Context, sql string, args ...any) ([]*auditLog, error) {
	rows, err := s.Pool.Query(ctx, sql, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var logs []*auditLog
	for rows.Next() {
		log, err := scanAuditLog(rows)
		if err != nil {
			return nil, err
		}
		logs = append(logs, log)
	}
	return logs, rows.Err()
}

func (h *handlers) writeAuditPage(ctx context.Context, w http.ResponseWriter, logs []*auditLog, total, limit, offset int64) {
	items, err := auditLogList(logs)
	if err != nil {
		h.internalError(ctx, w, "encode audit logs", err)
		return
	}
	out := pyjson.NewObject()
	out.Set("items", items)
	out.Set("total", total)
	out.Set("limit", limit)
	out.Set("offset", offset)
	policy.WriteModel(w, http.StatusOK, out, nil)
}

// listAuditLogs is audit_logs.py's list_audit_logs.
func (h *handlers) listAuditLogs(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	orgID, ok := adminOrgID(w, policy.UserFrom(ctx))
	if !ok {
		return
	}
	filter, limit, offset, errs := auditListQuery(r)
	if len(errs) > 0 {
		writeValidation(w, errs)
		return
	}
	if !h.requireFeature(ctx, w, auditLogFeature, orgID) {
		return
	}
	org, err := pythonparity.ParseUUID(orgID)
	if err != nil {
		policy.WriteInternal(w)
		return
	}
	conditions, args, valid := filter.where([]string{"org_id = $1"}, []any{org})
	if !valid || offset < 0 {
		policy.WriteInternal(w)
		return
	}
	logs, total, err := h.store.auditPage(ctx, conditions, args, limit, offset)
	if err != nil {
		h.internalError(ctx, w, "list audit logs", err)
		return
	}
	h.writeAuditPage(ctx, w, logs, total, limit, offset)
}

// listPlatformAuditLogs is audit_logs.py's list_platform_audit_logs.
func (h *handlers) listPlatformAuditLogs(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	filter, limit, offset, errs := auditListQuery(r)
	if len(errs) > 0 {
		writeValidation(w, errs)
		return
	}
	conditions, args, valid := filter.where(nil, nil)
	if !valid || offset < 0 {
		policy.WriteInternal(w)
		return
	}
	logs, total, err := h.store.auditPage(ctx, conditions, args, limit, offset)
	if err != nil {
		h.internalError(ctx, w, "list platform audit logs", err)
		return
	}
	h.writeAuditPage(ctx, w, logs, total, limit, offset)
}

// getAuditLog is audit_logs.py's get_audit_log.
func (h *handlers) getAuditLog(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	orgID, ok := adminOrgID(w, policy.UserFrom(ctx))
	if !ok {
		return
	}
	if !h.requireFeature(ctx, w, auditLogFeature, orgID) {
		return
	}
	org, orgErr := pythonparity.ParseUUID(orgID)
	logID, logOK := pathUUID(r, "log_id")
	if orgErr != nil || !logOK {
		policy.WriteInternal(w)
		return
	}
	log, err := scanAuditLog(h.store.Pool.QueryRow(ctx,
		`SELECT `+auditLogColumns+` FROM audit_logs WHERE id = $1 AND org_id = $2`, logID, org))
	if err != nil {
		h.internalError(ctx, w, "get audit log", err)
		return
	}
	if log == nil {
		policy.WriteDetail(w, http.StatusNotFound, "Audit log not found", nil)
		return
	}
	object, err := auditLogObject(log)
	if err != nil {
		h.internalError(ctx, w, "encode audit log", err)
		return
	}
	policy.WriteModel(w, http.StatusOK, object, nil)
}

// getResourceAuditHistory is audit_logs.py's get_resource_audit_history.
func (h *handlers) getResourceAuditHistory(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	orgID, ok := adminOrgID(w, policy.UserFrom(ctx))
	if !ok {
		return
	}
	var errs pybody.Errors
	limit := limitOnly(&errs, r.URL.Query(), 50, 500)
	if len(errs) > 0 {
		writeValidation(w, errs)
		return
	}
	if !h.requireFeature(ctx, w, auditLogFeature, orgID) {
		return
	}
	org, err := pythonparity.ParseUUID(orgID)
	if err != nil {
		policy.WriteInternal(w)
		return
	}
	logs, err := h.store.queryAuditLogs(ctx, `SELECT `+auditLogColumns+`
FROM audit_logs WHERE org_id = $1 AND resource_type = $2 AND resource_id = $3
ORDER BY created_at DESC LIMIT $4`, org, r.PathValue("resource_type"), r.PathValue("resource_id"), limit)
	if err != nil {
		h.internalError(ctx, w, "get resource audit history", err)
		return
	}
	h.writeAuditList(ctx, w, logs)
}

// getUserAuditActivity is audit_logs.py's get_user_audit_activity.
func (h *handlers) getUserAuditActivity(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	orgID, ok := adminOrgID(w, policy.UserFrom(ctx))
	if !ok {
		return
	}
	var errs pybody.Errors
	limit := limitOnly(&errs, r.URL.Query(), 50, 500)
	if len(errs) > 0 {
		writeValidation(w, errs)
		return
	}
	if !h.requireFeature(ctx, w, auditLogFeature, orgID) {
		return
	}
	org, orgErr := pythonparity.ParseUUID(orgID)
	userID, userOK := pathUUID(r, "user_id")
	if orgErr != nil || !userOK {
		policy.WriteInternal(w)
		return
	}
	logs, err := h.store.queryAuditLogs(ctx, `SELECT `+auditLogColumns+`
FROM audit_logs WHERE org_id = $1 AND user_id = $2 ORDER BY created_at DESC LIMIT $3`, org, userID, limit)
	if err != nil {
		h.internalError(ctx, w, "get user audit activity", err)
		return
	}
	h.writeAuditList(ctx, w, logs)
}

func (h *handlers) writeAuditList(ctx context.Context, w http.ResponseWriter, logs []*auditLog) {
	items, err := auditLogList(logs)
	if err != nil {
		h.internalError(ctx, w, "encode audit logs", err)
		return
	}
	policy.WriteModel(w, http.StatusOK, items, nil)
}
